File: agents/adapters/llm/src/llm_adapter/providers/openai.py
# SPDX-License-Identifier: Apache-2.0
"""OpenAI provider — streams token chunks via openai.AsyncOpenAI."""

from __future__ import annotations

from collections.abc import AsyncGenerator
from typing import TYPE_CHECKING

import openai
import structlog

if TYPE_CHECKING:
    from llm_adapter.config import OpenAIProviderConfig

log = structlog.get_logger()


def _build_messages(prompt: str, system: str | None) -> list[dict[str, str]]:
    """Assemble the messages list for the chat completions API.

    Args:
        prompt: User prompt text.
        system: Optional system instruction.

    Returns:
        Ordered list of role/content dicts.
    """
    msgs: list[dict[str, str]] = []
    if system:
        msgs.append({"role": "system", "content": system})
    msgs.append({"role": "user", "content": prompt})
    return msgs


async def stream_tokens(
    config: OpenAIProviderConfig,
    prompt: str,
    system: str | None,
    temperature: float,
    max_tokens: int,
) -> AsyncGenerator[bytes, None]:
    """Stream token chunks from the OpenAI chat completions API.

    Args:
        config: OpenAI provider configuration (API key + model).
        prompt: User prompt.
        system: Optional system message.
        temperature: Sampling temperature (0–2).
        max_tokens: Maximum token ceiling.

    Yields:
        UTF-8 encoded token chunk bytes.

    Raises:
        RuntimeError: When the API returns an error; mapped to UPSTREAM_ERROR by the handler.
    """
    client = openai.AsyncOpenAI(api_key=config.api_key.get_secret_value())
    try:
        stream = await client.chat.completions.create(
            model=config.model,
            messages=_build_messages(prompt, system),  # type: ignore[arg-type]
            temperature=temperature,
            max_tokens=max_tokens,
            stream=True,
        )
        async for chunk in stream:
            if chunk.choices and chunk.choices[0].delta.content:
                yield chunk.choices[0].delta.content.encode()
    except openai.OpenAIError as exc:
        log.warning("openai_upstream_error", error=str(exc)[:256])
        raise RuntimeError(str(exc)[:512]) from exc
    finally:
        await client.close()

File: agents/adapters/llm/internal/provider/openai.go
// SPDX-License-Identifier: Apache-2.0

package provider

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/zynax-io/zynax/agents/adapters/llm/internal/config"
)

// OpenAI Chat Completions streaming endpoint and SSE markers.
const (
	openAIDefaultBaseURL = "https://api.openai.com/v1"
	openAIChatPath       = "/chat/completions"
	sseDataPrefix        = "data: "
	sseDoneToken         = "[DONE]"
)

// openAIProvider streams token chunks from the OpenAI Chat Completions API over
// net/http using the Server-Sent-Events streaming protocol. The credential is
// held in a redacting Secret and only revealed onto the Authorization header at
// request time — never logged and never placed in an error message.
type openAIProvider struct {
	model     string
	maxTokens int
	baseURL   string
	secret    config.Secret
	client    *http.Client
}

// newOpenAIProvider binds OpenAI config + the resolved API key into a Provider.
func newOpenAIProvider(cfg config.ProviderConfig, secret config.Secret) *openAIProvider {
	return &openAIProvider{
		model:     cfg.Model,
		maxTokens: effectiveMaxTokens(cfg),
		baseURL:   openAIDefaultBaseURL,
		secret:    secret,
		client:    &http.Client{Timeout: 120 * time.Second},
	}
}

// openAIMessage is one role/content turn in the chat request.
type openAIMessage struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

// openAIRequest is the JSON body posted to the chat completions endpoint.
type openAIRequest struct {
	Model     string          `json:"model"`
	Messages  []openAIMessage `json:"messages"`
	MaxTokens int             `json:"max_tokens"`
	Stream    bool            `json:"stream"`
}

// openAIStreamChunk is one SSE frame from the streaming response.
type openAIStreamChunk struct {
	Choices []struct {
		Delta struct {
			Content string `json:"content"`
		} `json:"delta"`
	} `json:"choices"`
}

// Stream posts the prompt and relays each SSE delta.content as a Chunk.
func (p *openAIProvider) Stream(ctx context.Context, prompt string) (<-chan Chunk, error) {
	body, err := json.Marshal(openAIRequest{
		Model:     p.model,
		Messages:  []openAIMessage{{Role: "user", Content: prompt}},
		MaxTokens: p.maxTokens,
		Stream:    true,
	})
	if err != nil {
		return nil, fmt.Errorf("openai: marshal request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, p.baseURL+openAIChatPath, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("openai: build request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "text/event-stream")
	if !p.secret.IsZero() {
		req.Header.Set("Authorization", "Bearer "+p.secret.Reveal())
	}

	out := make(chan Chunk)
	go p.run(req, out)
	return out, nil
}

// run executes the request and pumps SSE frames into out, closing it when the
// stream ends or fails. Error messages never carry the response body or key.
func (p *openAIProvider) run(req *http.Request, out chan<- Chunk) {
	defer close(out)

	resp, err := p.client.Do(req) //nolint:gosec // URL host is operator-controlled config (provider base URL), not request input
	if err != nil {
		sendErr(out, fmt.Errorf("openai: %s", sanitiseErr(err.Error())))
		return
	}
	defer func() { _ = resp.Body.Close() }()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		sendErr(out, fmt.Errorf("openai: upstream returned HTTP %d", resp.StatusCode))
		return
	}

	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 0, 64*1024), 1024*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if !strings.HasPrefix(line, sseDataPrefix) {
			continue
		}
		payload := strings.TrimPrefix(line, sseDataPrefix)
		if payload == sseDoneToken {
			return
		}
		var frame openAIStreamChunk
		if err := json.Unmarshal([]byte(payload), &frame); err != nil {
			sendErr(out, fmt.Errorf("openai: decode stream frame: %s", sanitiseErr(err.Error())))
			return
		}
		if len(frame.Choices) > 0 && frame.Choices[0].Delta.Content != "" {
			if !send(req.Context(), out, Chunk{Text: frame.Choices[0].Delta.Content}) {
				return
			}
		}
	}
	if err := scanner.Err(); err != nil {
		sendErr(out, fmt.Errorf("openai: read stream: %s", sanitiseErr(err.Error())))
	}
}
